core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala

```diff
@@ -614,6 +585,74 @@ private[worker] class Worker(
     webUi.stop()
     metricsSystem.stop()
   }
+
+  private def trimFinishedExecutorsIfNecessary(): Unit = {
+    if (finishedExecutors.size > retainedExecutors) {
+      finishedExecutors.take(math.max(finishedExecutors.size / 10, 1)).foreach {
```

Comment:
The logic looks correct. I suppose you can take from the `keys()` rather than take pairs and match on them, but this is fine. 

Is this going to cause a `ConcurrentModificationException` though? you generally can't do this in Java and this is a JDK class. Does it work because take definitely makes a copy? Tests will tell us.

Is there in general a thread-safety problem here? I don't think the code tries to protect this data structure to begin with, so maybe it's not actually a problem.